File: clients/node/src/outcome.ts
// BeatOutcome tagged union, DropReason enum, BeatError, and the
// `classifySendError` function — mirrors Rust `BeatOutcome`,
// Python `BeatOutcome`/`DropReason`, and Go `BeatOutcome`/`DropReason`.

import {
  EAGAIN,
  ECONNREFUSED,
  ECONNRESET,
  ENOBUFS,
  ENOENT,
  ENOSPC,
  ENOTCONN,
  EPIPE,
  EWOULDBLOCK,
  errnoName,
} from "./errno.js";

// String-valued enum so the wire-side metrics label matches the Rust
// `Display` impl directly. The string values MUST stay byte-equal to
// the labels Prometheus sees from the observer (`varta_*_total{reason=…}`).
export enum DropReason {
  KernelQueueFull = "kernel queue full",
  NoObserver = "no observer",
  PeerGone = "peer gone",
  StorageFull = "storage full",
}

// Payload of `{ kind: "failed" }`. Extends `Error` so it can be thrown
// or logged like any other Node error, but carries structured `errno`
// and `kind` fields so callers don't have to parse `.message`.
export class BeatError extends Error {
  readonly errno: number;
  readonly kind: string;
  constructor(errno: number, kind: string, message?: string) {
    super(message ?? `${kind} (errno=${errno})`);
    this.name = "BeatError";
    this.errno = errno;
    this.kind = kind;
  }

  static fromNodeError(err: NodeJS.ErrnoException): BeatError {
    const code = typeof err.errno === "number" ? err.errno : 0;
    return new BeatError(code, code ? errnoName(code) : err.name || "Other", err.message);
  }
}

// Discriminated-union shape for results of `Varta.beat`. Mirrors the
// Rust algebraic enum and Python frozen dataclass.
export type BeatOutcome =
  | { readonly kind: "sent" }
  | { readonly kind: "dropped"; readonly reason: DropReason }
  | { readonly kind: "failed"; readonly error: BeatError };

export const BeatOutcomes = Object.freeze({
  sent(): BeatOutcome {
    return { kind: "sent" };
  },
  dropped(reason: DropReason): BeatOutcome {
    return { kind: "dropped", reason };
  },
  failed(error: BeatError): BeatOutcome {
    return { kind: "failed", error };
  },
});

export function isSent(o: BeatOutcome): boolean {
  return o.kind === "sent";
}
export function isDropped(o: BeatOutcome): boolean {
  return o.kind === "dropped";
}
export function isFailed(o: BeatOutcome): boolean {
  return o.kind === "failed";
}

// Translate a Node `send`-time error into a `BeatOutcome`.
//
// Mirrors `crates/varta-client/src/client.rs::classify_send_error` and
// the Python and Go equivalents. Exported because authors of custom
// `BeatTransport` implementations are likely to want the same bucketing.
export function classifySendError(err: NodeJS.ErrnoException): BeatOutcome {
  // Node's `ErrnoException.errno` is the negative POSIX value on some
  // calls (libuv convention) and the positive value on others. Take
  // the absolute value so the comparisons below work uniformly.
  const rawErrno = typeof err.errno === "number" ? err.errno : 0;
  const code = rawErrno < 0 ? -rawErrno : rawErrno;
  const sym = err.code ?? "";

  if (code === ENOBUFS || sym === "ENOBUFS") {
    return BeatOutcomes.dropped(DropReason.KernelQueueFull);
  }
  if (
    code === EAGAIN ||
    code === EWOULDBLOCK ||
    sym === "EAGAIN" ||
    sym === "EWOULDBLOCK"
  ) {
    return BeatOutcomes.dropped(DropReason.KernelQueueFull);
  }
  if (
    code === ECONNREFUSED ||
    code === ENOENT ||
    sym === "ECONNREFUSED" ||
    sym === "ENOENT"
  ) {
    return BeatOutcomes.dropped(DropReason.NoObserver);
  }
  if (
    code === ECONNRESET ||
    code === ENOTCONN ||
    code === EPIPE ||
    sym === "ECONNRESET" ||
    sym === "ENOTCONN" ||
    sym === "EPIPE"
  ) {
    return BeatOutcomes.dropped(DropReason.PeerGone);
  }
  if (code === ENOSPC || sym === "ENOSPC") {
    return BeatOutcomes.dropped(DropReason.StorageFull);
  }
  return BeatOutcomes.failed(BeatError.fromNodeError(err));
}

File: clients/go/outcome.go
package varta

import (
	"fmt"

	"github.com/aramirez087/Varta/clients/go/internal/errnoclass"
)

// DropReason identifies why a Beat returned Dropped. Mirrors the Rust
// enum and the Python StrEnum so the wire-side metrics label matches
// across languages.
type DropReason int

const (
	// KernelQueueFull — WouldBlock or ENOBUFS. Transient burst; the
	// observer is likely alive. Retry or rely on SetReconnectAfter.
	KernelQueueFull DropReason = iota + 1

	// NoObserver — NotFound or ConnectionRefused. The observer has not
	// bound the socket yet; expected during rolling restarts.
	NoObserver

	// PeerGone — ConnectionReset, NotConnected, or BrokenPipe. The
	// channel was live and disappeared (crash or shutdown). Call
	// Reconnect to recover.
	PeerGone

	// StorageFull — ENOSPC. Host filesystem full; operator
	// intervention required.
	StorageFull
)

// String returns the canonical wire-side label for the reason. Matches
// the Rust Display impl and the Python StrEnum values.
func (r DropReason) String() string {
	switch r {
	case KernelQueueFull:
		return "kernel queue full"
	case NoObserver:
		return "no observer"
	case PeerGone:
		return "peer gone"
	case StorageFull:
		return "storage full"
	}
	return "unknown"
}

// BeatError is the payload of BeatOutcome.Failed. Errno is the raw
// syscall errno (0 if not OS-derived); Kind is the symbolic name
// ("ENOENT", "EPERM", "Other", …).
type BeatError struct {
	Errno int
	Kind  string
}

// Error satisfies the error interface so callers can log the value
// directly.
func (e BeatError) Error() string {
	return fmt.Sprintf("varta: beat failed (errno=%d kind=%s)", e.Errno, e.Kind)
}

const (
	outcomeSent    = 1
	outcomeDropped = 2
	outcomeFailed  = 3
)

// BeatOutcome is the result of a single Beat call. Modeled as a tagged
// struct because Go lacks Rust's algebraic enums. Use the boolean
// IsSent/IsDropped/IsFailed predicates and the Reason/Error
// accessors; String renders a human-readable form.
type BeatOutcome struct {
	tag    uint8
	reason DropReason
	err    BeatError
}

// IsSent reports whether the kernel accepted the datagram.
func (o BeatOutcome) IsSent() bool { return o.tag == outcomeSent }

// IsDropped reports whether the datagram was not delivered.
func (o BeatOutcome) IsDropped() bool { return o.tag == outcomeDropped }

// IsFailed reports whether an unexpected I/O error occurred.
func (o BeatOutcome) IsFailed() bool { return o.tag == outcomeFailed }

// Reason returns the DropReason. Only valid when IsDropped is true.
func (o BeatOutcome) Reason() DropReason { return o.reason }

// Err returns the BeatError. Only valid when IsFailed is true.
func (o BeatOutcome) Err() BeatError { return o.err }

// String renders the outcome for logging / debugging.
func (o BeatOutcome) String() string {
	switch o.tag {
	case outcomeSent:
		return "sent"
	case outcomeDropped:
		return "dropped: " + o.reason.String()
	case outcomeFailed:
		return o.err.Error()
	}
	return "uninitialized"
}

// BeatOutcomeSent constructs a Sent outcome.
func BeatOutcomeSent() BeatOutcome { return BeatOutcome{tag: outcomeSent} }

// BeatOutcomeDropped constructs a Dropped outcome with the given reason.
func BeatOutcomeDropped(r DropReason) BeatOutcome {
	return BeatOutcome{tag: outcomeDropped, reason: r}
}

// BeatOutcomeFailed constructs a Failed outcome with the given error.
func BeatOutcomeFailed(e BeatError) BeatOutcome {
	return BeatOutcome{tag: outcomeFailed, err: e}
}

// ClassifySendError translates a transport-layer error (typically a
// *net.OpError wrapping a syscall.Errno) into a BeatOutcome. Exported
// so authors of custom transports can apply the same bucketing.
//
// Mirrors crates/varta-client/src/client.rs::classify_send_error and
// clients/python/src/varta/client.py::classify_send_error.
func ClassifySendError(err error) BeatOutcome {
	if err == nil {
		return BeatOutcomeSent()
	}
	switch errnoclass.Classify(err) {
	case errnoclass.BucketKernelQueueFull:
		return BeatOutcomeDropped(KernelQueueFull)
	case errnoclass.BucketNoObserver:
		return BeatOutcomeDropped(NoObserver)
	case errnoclass.BucketPeerGone:
		return BeatOutcomeDropped(PeerGone)
	case errnoclass.BucketStorageFull:
		return BeatOutcomeDropped(StorageFull)
	}
	return BeatOutcomeFailed(BeatError{
		Errno: int(errnoclass.Errno(err)),
		Kind:  errnoclass.Name(err),
	})
}
